### ISSUE TITLE
"This stream does not support writing" on Mono 5

### ISSUE THREAD

**Issue by sergey-tihon** [fsharp/FSharp.Data#1052]
I have a feeling that the line
```fsharp
JsonValue.Load("https://api.apis.guru/v2/list.json")
```
fails on Mono 5 from Mono `beta` channel `(Stable 5.0.0.94/1d0445b Wed May  3 17:40:24 UTC 2017)`

with error `System.InvalidOperationException: This stream does not support writing`

```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.TypeInitializationException: The type initializer for '<StartupCode$SwaggerProvider-Tests>.$APIsGuruFSCS' threw an exception. ---> System.TypeInitializationException: The type initializer for '<StartupCode$SwaggerProvider-Tests>.$APIsGuru' threw an exception. ---> System.InvalidOperationException: This stream does not support writing
  at System.IO.Compression.DeflateStream.BeginWrite (System.Byte[] buffer, System.Int32 offset, System.Int32 count, System.AsyncCallback cback, System.Object state) [0x00021] in <fc00127690464f449eb78053a97d2150>:0 
  at System.IO.Stream+<>c.<BeginEndWriteAsync>b__53_0 (System.IO.Stream stream, System.IO.Stream+ReadWriteParameters args, System.AsyncCallback callback, System.Object state) [0x00000] in <0f03d209a56046bbb0cf1e18c77d3be4>:0 
  at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncTrim[TInstance,TArgs] (TInstance thisRef, TArgs args, System.Func`5[T1,T2,T3,T4,TResult] beginMethod, System.Func`3[T1,T2,TResult] endMethod) [0x00008] in <0f03d209a56046bbb0cf1e18c77d3be4>:0 
  at System.IO.Stream.BeginEndWriteAsync (System.Byte[] buffer, System.Int32 offset, System.Int32 count) [0x00000] in <0f03d209a56046bbb0cf1e18c77d3be4>:0 
  at System.IO.Stream.WriteAsync (System.Byte[] buffer, System.Int32 offset, System.Int32 count, System.Threading.CancellationToken cancellationToken) [0x00009] in <0f03d209a56046bbb0cf1e18c77d3be4>:0 
  at System.IO.Compression.GZipStream.WriteAsync (System.Byte[] array, System.Int32 offset, System.Int32 count, System.Threading.CancellationToken cancellationToken) [0x00006] in <fc00127690464f449eb78053a97d2150>:0 
  at System.IO.Compression.GZipStream.BeginRead (System.Byte[] array, System.Int32 offset, System.Int32 count, System.AsyncCallback asyncCallback, System.Object asyncState) [0x00009] in <fc00127690464f449eb78053a97d2150>:0 
  at Microsoft.FSharp.Control.CommonExtensions+AsyncRead@2134.Invoke (System.Tuple`5[T1,T2,T3,T4,T5] tupledArg) [0x00025] in <5893d081904cf4daa745038381d09358>:0 
  at <StartupCode$FSharp-Core>.$Control+FromBeginEnd@1897-4[TArg1,TArg2,TArg3].Invoke (System.Tuple`2[T1,T2] tupledArg) [0x0002e] in <5893d081904cf4daa745038381d09358>:0 
  at <StartupCode$FSharp-Core>.$Control+FromBeginEnd@1836-1[T].Invoke (System.Threading.CancellationToken _arg4) [0x0005c] in <5893d081904cf4daa745038381d09358>:0 
  at Microsoft.FSharp.Control.AsyncBuilderImpl+args@825-1[a,b].Invoke (a a) [0x00016] in <5893d081904cf4daa745038381d09358>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <0f03d209a56046bbb0cf1e18c77d3be4>:0 
  at Microsoft.FSharp.Control.AsyncBuilderImpl.commit[a] (Microsoft.FSharp.Control.AsyncBuilderImpl+AsyncImplResult`1[T] res) [0x0002d] in <5893d081904cf4daa745038381d09358>:0 
  at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronouslyInCurrentThread[a] (System.Threading.CancellationToken token, Microsoft.FSharp.Control.FSharpAsync`1[T] computation) [0x00029] in <5893d081904cf4daa745038381d09358>:0 
  at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronously[a] (System.Threading.CancellationToken token, Microsoft.FSharp.Control.FSharpAsync`1[T] computation, Microsoft.FSharp.Core.FSharpOption`1[T] timeout) [0x00014] in <5893d081904cf4daa745038381d09358>:0 
  at Microsoft.FSharp.Control.FSharpAsync.RunSynchronously[T] (Microsoft.FSharp.Control.FSharpAsync`1[T] computation, Microsoft.FSharp.Core.FSharpOption`1[T] timeout, Microsoft.FSharp.Core.FSharpOption`1[T] cancellationToken) [0x00071] in <5893d081904cf4daa745038381d09358>:0 
  at FSharp.Data.JsonValue.Load (System.String uri, Microsoft.FSharp.Core.FSharpOption`1[T] cultureInfo) [0x00009] in <591304705732e8dca745038370041359>:0 
```

Issue found here https://github.com/fsprojects/SwaggerProvider/pull/76

**Update 1** Seems that `Http.RequestString` also fail

**Comment by sergey-tihon** [comment]
Issue is reproduced by @nosami [here](https://github.com/fsprojects/SwaggerProvider/pull/76#issuecomment-300517554)

```
 F# Interactive for F# 4.1
 Freely distributed under the Apache 2.0 Open Source License
 For help type #help;;

#r "/Users/jason/src/FSharp.Data/bin/FSharp.Data.dll";;
 --> Referenced '/Users/jason/src/FSharp.Data/bin/FSharp.Data.dll' (file may be locked by F# Interactive process)

open FSharp.Data;;

JsonValue.Load("https://api.apis.guru/v2/list.json");;
 System.InvalidOperationException: This stream does not support writing
   at System.IO.Compression.DeflateStream.BeginWrite (System.Byte[] buffer, System.Int32 offset, System.Int32 count, System.AsyncCallback cback, System.Object state) [0x00021] in <f666904f99e949f49991eeadc1f953e0>:0 
   at System.IO.Stream+<>c.<BeginEndWriteAsync>b__53_0 (System.IO.Stream stream, System.IO.Stream+ReadWriteParameters args, System.AsyncCallback callback, System.Object state) [0x00000] in <b697cffb61b74023aa9c587e6c49beb3>:0 
   at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncTrim[TInstance,TArgs] (TInstance thisRef, TArgs args, System.Func`5[T1,T2,T3,T4,TResult] beginMethod, System.Func`3[T1,T2,TResult] endMethod) [0x00008] in <b697cffb61b74023aa9c587e6c49beb3>:0 
   at System.IO.Stream.BeginEndWriteAsync (System.Byte[] buffer, System.Int32 offset, System.Int32 count) [0x00000] in <b697cffb61b74023aa9c587e6c49beb3>:0 
   at System.IO.Stream.WriteAsync (System.Byte[] buffer, System.Int32 offset, System.Int32 count, System.Threading.CancellationToken cancellationToken) [0x00009] in <b697cffb61b74023aa9c587e6c49beb3>:0 
   at System.IO.Compression.GZipStream.WriteAsync (System.Byte[] array, System.Int32 offset, System.Int32 count, System.Threading.CancellationToken cancellationToken) [0x00006] in <f666904f99e949f49991eeadc1f953e0>:0 
   at System.IO.Compression.GZipStream.BeginRead (System.Byte[] array, System.Int32 offset, System.Int32 count, System.AsyncCallback asyncCallback, System.Object asyncState) [0x00009] in <f666904f99e949f49991eeadc1f953e0>:0 
   at Microsoft.FSharp.Control.CommonExtensions+AsyncRead@2134.Invoke (System.Tuple`5[T1,T2,T3,T4,T5] tupledArg) [0x00025] in <5893d081904cf4daa745038381d09358>:0 
   at <StartupCode$FSharp-Core>.$Control+FromBeginEnd@1897-4[TArg1,TArg2,TArg3].Invoke (System.Tuple`2[T1,T2] tupledArg) [0x0002e] in <5893d081904cf4daa745038381d09358>:0 
   at <StartupCode$FSharp-Core>.$Control+FromBeginEnd@1836-1[T].Invoke (System.Threading.CancellationToken _arg4) [0x0005c] in <5893d081904cf4daa745038381d09358>:0 
   at Microsoft.FSharp.Control.AsyncBuilderImpl+args@825-1[a,b].Invoke (a a) [0x00016] in <5893d081904cf4daa745038381d09358>:0 
 --- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <b697cffb61b74023aa9c587e6c49beb3>:0 
   at Microsoft.FSharp.Control.AsyncBuilderImpl.commit[a] (Microsoft.FSharp.Control.AsyncBuilderImpl+AsyncImplResult`1[T] res) [0x0002d] in <5893d081904cf4daa745038381d09358>:0 
   at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronouslyInCurrentThread[a] (System.Threading.CancellationToken token, Microsoft.FSharp.Control.FSharpAsync`1[T] computation) [0x00029] in <5893d081904cf4daa745038381d09358>:0 
   at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronously[a] (System.Threading.CancellationToken token, Microsoft.FSharp.Control.FSharpAsync`1[T] computation, Microsoft.FSharp.Core.FSharpOption`1[T] timeout) [0x00014] in <5893d081904cf4daa745038381d09358>:0 
   at Microsoft.FSharp.Control.FSharpAsync.RunSynchronously[T] (Microsoft.FSharp.Control.FSharpAsync`1[T] computation, Microsoft.FSharp.Core.FSharpOption`1[T] timeout, Microsoft.FSharp.Core.FSharpOption`1[T] cancellationToken) [0x00071] in <5893d081904cf4daa745038381d09358>:0 
   at FSharp.Data.JsonValue.Load (System.String uri, Microsoft.FSharp.Core.FSharpOption`1[T] cultureInfo) [0x00009] in <57286beabdc81bcda7450383ea6b2857>:0 
   at <StartupCode$FSI_0002>.$FSI_0002.main@ () [0x00001] in <0258e23fb17048d59131c566e6f8f494>:0 
   at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
   at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in <b697cffb61b74023aa9c587e6c49beb3>:0 

```

**Comment by mateuszlewko** [comment]
I'm having exactly the same issue, so I guess it happens to everyone.

**Comment by sergey-tihon** [comment]
Update, seems that Mono does not has it's own [System.IO.Compression](https://github.com/mono/mono/blob/master/mcs/class/System.IO.Compression/System.IO.Compression.dll.sources) and reuse implementation from `corefx`!

So, the bug probably should be submitted to dotnet core team (if someone will repro it on .net core).
This issue may be relevant ... https://github.com/dotnet/corefx/issues/11497


P.S. mono issue https://bugzilla.xamarin.com/show_bug.cgi?id=56260

**Comment by easysoft2k15** [comment]
@sergey-tihom, did You manage to disable *gzip*?
I thought I could write:

`req.AutomaticDecompression <- DecompressionMethods.None`

but seems like `AutomaticDecompression` property is not available on `Profile7`.

**Comment by sergey-tihon** [comment]
Seems that issue fixed in released Mono 5.0.1
